### ISSUE TITLE
Index build procedure misses tuples inserted by transactions started before DDL

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#3458]
Reproduced by Travis CI:
```
vinyl/ddl.test.lua                                              [ fail ]
Test failed! Result content mismatch:
--- vinyl/ddl.result	Tue Jun 12 17:53:32 2018
+++ vinyl/ddl.reject	Tue Jun 12 18:00:56 2018
@@ -701,7 +701,7 @@
 ...
 box.space.test.index.pk:count() == box.space.test.index.sk:count()
 ---
-- true
+- false
 ...
 box.space.test.index.pk:count() == box.space.test.index.tk:count()
 ---
@@ -710,7 +710,7 @@
 test_run:cmd("restart server default")
 box.space.test.index.pk:count() == box.space.test.index.sk:count()
 ---
-- true
+- false
 ...
 box.space.test.index.pk:count() == box.space.test.index.tk:count()
 ---
@@ -722,7 +722,7 @@
 ...
 box.space.test.index.pk:count() == box.space.test.index.sk:count()
 ---
-- true
+- false
 ...
 box.space.test.index.pk:count() == box.space.test.index.tk:count()
 ---
```

The bug seems to occur, because when we install [on_replace trigger](https://github.com/tarantool/tarantool/blob/5a772639aa05479fa91ec696596f9cb921b6fff4/src/box/vinyl.c#L4359) that is supposed to propagate changes done to the altered space while the new index is being built there may be an active transaction with a non-empty write set. If this transaction gets committed while the build is in progress (it can't get committed after the space has been altered, because there will be tuple format mismatch then), changes done by it may be missed by the build procedure (it could advance the cursor beyond that point by then), neither will they be propagated by the on_replace trigger (the trigger was installed after the changes were executed). As a result, some records will be present in the primary index, but missing in the newly built index.

The problem is demonstrated by the script below:
```
yaml = require('yaml')
fiber = require('fiber')

box.cfg{}

s = box.schema.space.create('test', {engine = 'vinyl'})
s:create_index('pk', {page_size = 16})
for i = 1, 10 do
    s:replace{i, i, string.rep('x', 10)}
end
box.snapshot()

box.error.injection.set('ERRINJ_VY_READ_PAGE_TIMEOUT', 0.1)

c = fiber.channel(1)
fiber.create(function()
    box.begin()
    s:replace{100, 1} -- violates unique constraint of the new index
    fiber.sleep(0.2)
    box.commit()
    c:put(true)
end)

s:create_index('sk', {unique = true, parts = {2, 'unsigned'}}) -- should fail as per above, but doesn't
c:get()

assert(s.index.sk:count() == s.index.pk:count()) -- this assertion fails
os.exit(0)
```

If you replace `s:replace{100, 1}` with `s:replace{100}` you'll get an uncaught format mismatch resulting in a crash:
```
tarantool: /home/vlad/src/tarantool/src/box/vy_stmt.c:240: vy_stmt_new_with_ops: Assertion `field_count >= format->min_field_count' failed.
```
We should probably flush all pending transactions before installing the on_replace trigger.

**Comment by locker** [comment]
To make the issue nearly 100% reproducible, apply the following patch:
```
diff --git a/test/vinyl/ddl.result b/test/vinyl/ddl.result
index 3e65e232..bb4fc984 100644
--- a/test/vinyl/ddl.result
+++ b/test/vinyl/ddl.result
@@ -656,7 +656,6 @@ last_val = 1000;
 ---
 ...
 function gen_load()
-    fiber.sleep(0.001)
     local s = box.space.test
     for i = 1, 200 do
         local op = math.random(4)
diff --git a/test/vinyl/ddl.test.lua b/test/vinyl/ddl.test.lua
index 45c5cf8e..f7bf03bb 100644
--- a/test/vinyl/ddl.test.lua
+++ b/test/vinyl/ddl.test.lua
@@ -253,7 +253,6 @@ box.commit();
 last_val = 1000;

 function gen_load()
-    fiber.sleep(0.001)
     local s = box.space.test
     for i = 1, 200 do
         local op = math.random(4)
```
(Actually, I put that sleep there just to make the test pass :-)

**Comment by locker** [comment]
Fixed by c602cc2c82963ca34d71a55e94c36306b854460e
